src/goto-programs/goto_program_template.h

```diff
@@ -80,60 +79,129 @@ class goto_program_templatet
     return *this;
   }
 
-  /*! \brief Container for an instruction of the goto-program
-  */
+  /// This class represents a node in the control flow graph of the function.
+  /// The successor nodes are determined by:
+  ///
+  /// - The position of this node in the field
+  ///   `goto_program_templatet::instructions` of the external class. The
+  ///   instruction stored after this instruction is implicitly a successor of
+  ///   this instruction (unless this instruction unconditionally jumps
+  ///   elsewhere).
+  /// - The field `targets` in this class, storing a list of pointers
+  ///   (usually only 1) to other instructions in the CFG.
+  ///
+  /// Three other fields are key:
+  ///
+  /// - type:  an enum value describing the action performed by this instruction
+  /// - guard: an (arbitrarily complex) expression (usually an \ref exprt)
+  /// - code:  a code statement (usually a \ref codet)
+  ///
+  /// The meaning of an instruction node depends on the `type` field. Different
+  /// kinds of instructions make use of the fields `guard` and `code` for
+  /// different purposes.  We list below, using a mixture of pseudo code and
+  /// plain English, the meaning of different kinds of instructions.
+  /// We use `guard`, `code`, and `targets` to mean the value of the
+  /// respective fields in this class:
```

Comment:
I'm not sure the semantics of internal of a formal-verification tool should be described in a completely informal way. It may be good to refer to Hoare logic.